### PR TITLE
cmd/{k8s-operator,containerboot},k8s-operator,kube: reconcile ExternalName Services for ProxyGroup

### PR DESCRIPTION
**Pull request by irbekrm**
This PR is the second part for https://github.com/tailscale/tailscale/issues/13406.

It adds two new reconcile loops to the Kubernetes operator that:
- reconciles user created ExternalName Services that specify a tailnet target and a ProxyGroup on which to expose the target to cluster workloads
- for each ExternalName service, update the config that's mounted to the ProxyGroup proxies. Proxies read the config and update firewall rules accordingly, see https://github.com/tailscale/tailscale/pull/13531
- creates a ClusterIP Service for routing cluster traffic from the ExternalName Service
- creates an EndpointSlice with endpoints for the ClusterIP Service ^ that gets dynamically updated with proxy Pod IPs that are currently ready to route traffic

### CODE REVIEW

File: cmd/k8s-operator/egress-services.go [hunk 1]
```diff
@@ -0,0 +1,714 @@
...
+func isEgressSvcForProxyGroup(obj client.Object) bool {
+	s, ok := obj.(*corev1.Service)
+	if !ok {
+		return false
+	}
+	annots := s.ObjectMeta.Annotations
+	return annots[AnnotationProxyGroup] != "" && (annots[AnnotationTailnetTargetFQDN] != "" || annots[AnnotationTailnetTargetIP] != "")
+}
+
+func egressSvcConfig(ctx context.Context, cl client.Client, tailnetSvc, proxyGroupName, tsNamespace string) (*corev1.ConfigMap, *egressservices.Configs, *egressservices.Config, error) {
```
[the remark this answers is not in the document]
Comment (tomhjp) [reply cmd/k8s-operator/egress-services.go:605]: Also it seems like it would be cleaner to just return one of `Configs` or `Config`, because the `Config` value is simply a subset of the `Configs` value.

File: cmd/k8s-operator/egress-eps.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+		return res, nil
+	}
+
+	// Check which Pods in ProxyGroup are ready to route traffic to this
+	// egress service.
+	podList := &corev1.PodList{}
+	if err := er.List(ctx, podList, client.MatchingLabels(map[string]string{labelProxyGroup: proxyGroupName})); err != nil {
+		return res, fmt.Errorf("error listing Pods for ProxyGroup %s: %w", proxyGroupName, err)
+	}
+	if len(podList.Items) == 0 {
```
Comment (tomhjp) [inline cmd/k8s-operator/egress-eps.go:103]: I don't think we should return early here. What if the number of pods is newly 0 and we don't update the endpoint slice as a result?

**Review by tomhjp — COMMENTED** [review]
Looking good! I still have `egress-services.go` to review, but wanted to post what I have so far to unblock as much as possible

File: cmd/k8s-operator/operator.go [hunk 3]
```diff
@@ -712,3 +757,105 @@ func isMagicDNSName(name string) bool {
...
+		}
+		return reqs
+	}
+}
+
+func egressSvcsFromEgressProxyGroup(cl client.Client, logger *zap.SugaredLogger) handler.MapFunc {
+	return func(_ context.Context, o client.Object) []reconcile.Request {
+		pg, ok := o.(*tsapi.ProxyGroup)
+		if !ok {
+			logger.Infof("[unexpected] ProxyGroup handler triggered for an object that is not a ProxyGroup")
```
[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/k8s-operator/operator.go:832]: `logger.Error` prints a stack trace to the c/r's caller of this reconcile that, in this case is not valuable and makes it actually harder to read the logs

File: cmd/k8s-operator/egress-eps.go (hunk 2 above)
Comment (irbekrm) [reply cmd/k8s-operator/egress-eps.go:103]: Good catch, thanks

File: cmd/k8s-operator/operator.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomhjp) [reply cmd/k8s-operator/operator.go:832]: Oh good point 👍

File: k8s-operator/conditions.go [hunk 4]
```diff
@@ -116,6 +128,33 @@ func ProxyClassIsReady(pc *tsapi.ProxyClass) bool {
...
+		idx := xslices.IndexFunc(svc.Status.Conditions, func(cond metav1.Condition) bool {
+			return cond.Type == string(typ)
+		})
+		if idx == -1 {
+			return false
+		}
+		cond := svc.Status.Conditions[idx]
+		if cond.Status != metav1.ConditionTrue {
+			return false
+		}
```
[the remark this answers is not in the document]
Comment (irbekrm) [reply k8s-operator/conditions.go:153]: nice

File: cmd/k8s-operator/svc.go [hunk 5]
```diff
@@ -354,6 +354,10 @@ func validateService(svc *corev1.Service) []string {
 			violations = append(violations, fmt.Sprintf("invalid value of annotation %s: %q does not appear to be a valid MagicDNS name", AnnotationTailnetTargetFQDN, fqdn))
 		}
 	}
+
+	// TODO(irbekrm): validate that tailscale.com/tailnet-ip annotation is a
```
[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/k8s-operator/svc.go:358]: https://github.com/tailscale/tailscale/issues/13671

File: cmd/k8s-operator/egress-services.go [hunk 6]
```diff
@@ -0,0 +1,714 @@
...
+	}
+	return cm, cfgs, nil, nil
+}
+
+// egressSvcChildResourceLabels returns labels that should be applied to the
+// ClusterIP Service and the EndpointSlice created for the egress service.
+// TODO(irbekrm): we currently set a bunch of labels based on Kubernetes
+// resource names (ProxyGroup, Service). Maximum allowed label length is 63
+// chars whilst the maximum allowed resource name length is 253 chars, so we
+// should probably validate and trunkate (?) the names is they are too long.
```
Comment (tomhjp) [inline cmd/k8s-operator/egress-services.go:633]: This is a good point, might be a good one to fix before promoting CRDs to beta.

```suggestion
// should probably validate and truncate (?) the names is they are too long.
```

File: cmd/k8s-operator/egress-services_test.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+				{
+					Name:     "https",
+					Protocol: "TCP",
+					Port:     443,
+				},
+			},
+		},
+	}
+
+	// 1. Create an ExternalName Service for a ProxyGroup that is not ready yet.
```
[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/k8s-operator/egress-services_test.go:94]: This test shows a basic flow of an ExternalName Service being created and deleted, I am not planning on expanding this one, but also would rather keep it so that it follows the reconciler flow (a resource is not ready, then the same resource becomes ready etc) than spit it into separate cases.

We can add individual tests that test more edge cases for individual actions and could be structured differently. But also, for reconciler tests, I do prefer tests that iterate on the same resource as it would happen in cluster

File: cmd/k8s-operator/egress-services.go [hunk 8]
```diff
@@ -0,0 +1,714 @@
...
+	violations := validateService(svc)
+	if len(svc.Spec.Ports) == 0 {
+		violations = append(violations, "egress Service for ProxyGroup must have at least one target Port specified")
+	}
+	if svc.Spec.Type != corev1.ServiceTypeExternalName {
+		violations = append(violations, fmt.Sprintf("unexpected egress Service type %s. The only supported type is ExternalName.", svc.Spec.Type))
+	}
+	if pg.Spec.Type != tsapi.ProxyGroupTypeEgress {
+		violations = append(violations, fmt.Sprintf("egress Service references ProxyGroup of type %s, must be type %s", pg.Spec.Type, tsapi.ProxyGroupTypeEgress))
+	}
```
Comment (tomhjp) [inline cmd/k8s-operator/egress-services.go:528]: Maybe add another check for if they've set exactly one of target FQDN and IP?

File: cmd/k8s-operator/egress-services.go [hunk 9]
```diff
@@ -0,0 +1,714 @@
...
+		base = base + "-"                // re-instate the dash
+	}
+}
+
+// unusedPort returns a port in range [3000 - 4000). The caller must ensure that
+// usedPorts does not contain all ports in range [3000 - 4000).
+func unusedPort(usedPorts sets.Set[int32]) int32 {
+	foundFreePort := false
+	var suggestPort int32
+	for !foundFreePort {
```
Comment (tomhjp) [inline cmd/k8s-operator/egress-services.go:559]: I'm sure it's probably fine, but this algorithm's runtime will get a little bit choppy when there aren't many ports left, e.g. consider how lucky we need to get when there's only 1 remaining port. The expected number of iterations is 1000, but it's technically unbounded. Any reason not to just loop through all valid ports until we find one not in `usedPorts`? I don't see any downside as long as we can guarantee only one control loop is running at any one time (which we can for a specific reconcile target right?).

If we cared to do better than that, we could track the port number we got up to and start there next time. That would mean 1 iteration the first 1000 times, and then always starting with the port we allocated longest ago from then on. However, I don't think it's worth adding that complexity for the number of ports right now.

File: cmd/k8s-operator/egress-eps.go [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+	svcStatusBS := stateS.Data[egressservices.KeyEgressServices]
+	if len(svcStatusBS) == 0 {
+		l.Debugf("proxy's state Secret does not contain egress services status, waiting...")
+		return false, nil
+	}
+	svcStatus := &egressservices.Status{}
+	if err := json.Unmarshal(svcStatusBS, svcStatus); err != nil {
+		return false, fmt.Errorf("error unmarshalling egress service status: %w", err)
+	}
+	if !strings.EqualFold(podIP, svcStatus.PodIP) {
```
[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/k8s-operator/egress-eps.go:169]: I was thinking about this earlier.

The Pod IP in status is nicer for readability when debugging (because we operate on IPs anyaway)

Given that:
- ProxyGroup Pods are recreated one by one if needs be
- a Pod is tied to the same state Secret with the status with IP

It seems like the only case where a clash could happen would be if a Pod 0 is recreated and gets the same IP assigned when it comes up. As far as I have seen CNIs don't allocate the same IP to a recreated Pod, usually it's something like N+1 for the latest one used. So, this seems extremely unlikely unless that cluster only has a single Pod IP left to allocate.

So to me it seems that Pod IP is a good enough identifier because we really only need that IP to distinguish this deployment of Pod 0 from a previous deployment of Pod 0, not to uniquely identify this Pod amongst other replicas in ProxyGroup or amongst other Pods in cluster

**Comment by irbekrm** [comment]
Thank you for the review @tomhjp I think I have now addressed all the comments

File: cmd/k8s-operator/egress-services.go [hunk 11]
```diff
@@ -0,0 +1,714 @@
...
+			return base
+		}
+		base = base[:len(base)-1-excess] // cut off the excess chars
+		base = base + "-"                // re-instate the dash
+	}
+}
+
+// unusedPort returns a port in range [3000 - 4000). The caller must ensure that
+// usedPorts does not contain all ports in range [3000 - 4000).
+func unusedPort(usedPorts sets.Set[int32]) int32 {
```
Comment (tomhjp) [inline cmd/k8s-operator/egress-services.go:556]: Did you mean to use the apimachinery set rather than this repo's set package?

File: cmd/k8s-operator/egress-services.go [hunk 12]
```diff
@@ -0,0 +1,714 @@
...
+		return fmt.Errorf("error deleting egress service config: %w", err)
+	}
+	return nil
+}
+
+// usedPortsForPG calculates the currently used match ports for ProxyGroup
+// containers. It does that by looking by retrieving all target ports of all
+// ClusterIP Services created for egress services exposed on this ProxyGroup's
+// proxies.
+// TODO(irbekrm): this is currently good enough because we only have a single worker and
```
Comment (tomhjp) [inline cmd/k8s-operator/egress-services.go:404]: Thanks for writing this down - super helpful comment!

File: cmd/k8s-operator/egress-services.go [hunk 13]
```diff
@@ -0,0 +1,714 @@
...
+	// assume that the egress ExternalName Services are always created for
+	// Tailscale operator specifically.
+	if !svc.DeletionTimestamp.IsZero() {
+		l.Info("Service is being deleted, ensuring resource cleanup")
+		return res, esr.maybeCleanup(ctx, svc, l)
+	}
+
+	oldStatus := svc.Status.DeepCopy()
+	defer func() {
+		l.Debugf("looking at whether to update Service")
```
Comment (tomhjp) [inline cmd/k8s-operator/egress-services.go:132]: Probably a redundant log line?

File: cmd/k8s-operator/egress-services.go [hunk 14]
```diff
@@ -0,0 +1,714 @@
...
+	l.Infof("updating configuration...")
+	usedPorts, err := esr.usedPortsForPG(ctx, proxyGroupName)
+	if err != nil {
+		return nil, false, fmt.Errorf("error calculating used ports for ProxyGroup %s: %w", proxyGroupName, err)
+	}
+
+	oldClusterIPSvc := clusterIPSvc.DeepCopy()
+	// loop over ClusterIP Service ports, remove any that are not needed.
+	i := 0
+	for {
```
Comment (tomhjp) [inline cmd/k8s-operator/egress-services.go:232]: A slightly simpler way to handle looping over a slice that you're deleting from as you go is to iterate backwards: https://go.dev/play/p/ZTRGMnlyE-N

File: cmd/k8s-operator/egress-services.go [hunk 15]
```diff
@@ -0,0 +1,714 @@
...
+		found := false
+		for _, wantsPM := range svc.Spec.Ports {
+			if wantsPM.Port == pm.Port && strings.EqualFold(string(wantsPM.Protocol), string(pm.Protocol)) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			l.Debugf("portmapping %s:%d -> %s:%d is no longer required, removing", pm.Protocol, pm.TargetPort.IntVal, pm.Protocol, pm.Port)
+			clusterIPSvc.Spec.Ports = append(clusterIPSvc.Spec.Ports[:i], clusterIPSvc.Spec.Ports[i+1:]...)
```
Comment (tomhjp) [inline cmd/k8s-operator/egress-services.go:246]:
```suggestion
			clusterIPSvc.Spec.Ports = slices.Delete(clusterIPSvc.Spec.Ports, i, i+1:])
```

File: cmd/k8s-operator/egress-services.go [hunk 16]
```diff
@@ -0,0 +1,714 @@
...
+		if clusterIPSvc, err = createOrUpdate(ctx, esr.Client, esr.tsNamespace, clusterIPSvc, func(svc *corev1.Service) {
+			svc.Labels = clusterIPSvc.Labels
+			svc.Spec = clusterIPSvc.Spec
+		}); err != nil {
+			return nil, false, fmt.Errorf("error ensuring ClusterIP Service: %v", err)
+		}
+	}
+
+	crl := egressSvcChildResourceLabels(svc)
+	// TODO(irbekrm): support IPv6 (but need to investigate how kube proxy
```
Comment (tomhjp) [inline cmd/k8s-operator/egress-services.go:292]: nit: unclosed parenthesis

```suggestion
	// TODO(irbekrm): support IPv6, but need to investigate how kube proxy
```

File: cmd/k8s-operator/egress-services_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (tomhjp) [reply cmd/k8s-operator/egress-services_test.go:94]: Oh absolutely agreed, I only meant to sprinkle in some additional `t.Run(...` calls within this existing test much like in [TestRecorder](https://github.com/tailscale/tailscale/blob/main/cmd/k8s-operator/tsrecorder_test.go#L29)

**Comment by tomhjp** [comment]
Ah sorry I hadn't seen you'd already addressed the previous comments - will take another look after lunch

File: cmd/k8s-operator/egress-services.go (hunk 6 above)
Comment (irbekrm) [reply cmd/k8s-operator/egress-services.go:633]: > trunkate

could argue that it's a suitable word for kube

File: cmd/k8s-operator/egress-services.go [hunk 17]
```diff
@@ -0,0 +1,714 @@
...
+	r := fmt.Sprintf("ProxyGroup:%s", svc.Annotations[AnnotationProxyGroup])
+	tt := tailnetTargetFromSvc(svc)
+	s := cfg{
+		Ports:         svc.Spec.Ports,
+		TailnetTarget: tt,
+		ProxyGroup:    svc.Annotations[AnnotationProxyGroup],
+	}
+	h := cfgHash(s, l)
+	r = fmt.Sprintf("%s:Config:%s", r, h)
+	if configured {
```
[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/k8s-operator/egress-services.go:688]: I can pull it up to make it easier to read sequentially

Not sure if there is a value in using string builder- the memory consumption is small and the builder will introduce a bunch of error checks

File: cmd/k8s-operator/egress-services.go (hunk 9 above)
Comment (irbekrm) [reply cmd/k8s-operator/egress-services.go:559]: > we could track the port number we got up to and start there next time.

Yes, I was thinking that when/if we will anyway need to store used ports on ProxyGroup status, we could also store the last used port there

> I'm sure it's probably fine, but this algorithm's runtime will get a little bit choppy when there aren't many ports left, e.g. consider how lucky we need to get when there's only 1 remaining port

Yeah, really we should figure out what the limits are (number of iptables rules, configmap size etc).
I generally prefer to restrict everything on start and loosen later if needed, so that we get more freedom to change things if needed without breaking anyone. For example, what if for some reasons users want to introduce the ability to define port ranges for network policy config reasons or if we need to set aside a port range for something else - it'd be nice to have a port range free for that
I can bump the 1000 up to something like 10000 though, that should probably make the possibility of running out of ports very unlikely + we can improve the algorithm once we set used ports on ProxyGroup status

File: cmd/k8s-operator/egress-services_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/k8s-operator/egress-services_test.go:94]: Oh, that's nice, thanks

File: cmd/k8s-operator/operator.go [hunk 18]
```diff
@@ -378,13 +378,15 @@ func runReconcilers(opts reconcilerOpts) {
...
 	podsSecretsFilter := handler.EnqueueRequestsFromMapFunc(egressEpsFromEgressPGChildResources(mgr.GetClient(), opts.log, opts.tailscaleNamespace))
+	epsFromExtNSvcFilter := handler.EnqueueRequestsFromMapFunc(epsFromExternalNameService(mgr.GetClient(), opts.log))
 
 	err = builder.
 		ControllerManagedBy(mgr).
 		Named("egress-eps-reconciler").
 		Watches(&discoveryv1.EndpointSlice{}, epsFilter).
 		Watches(&corev1.Pod{}, podsSecretsFilter).
 		Watches(&corev1.Secret{}, podsSecretsFilter).
+		Watches(&corev1.Service{}, epsFromExtNSvcFilter).
```
Comment (irbekrm) [inline cmd/k8s-operator/operator.go:389]: Additionally had to add this watch as the EndpointSlice reconciler checks for this service to have Valid and Configured conditions before proceeding

**Comment by irbekrm** [comment]
Thank you for the review @tomhjp, I believe I have now addressed all the comments

**Review by tomhjp — APPROVED** [review]
LGTM! 🎉
